_posts/2025-02-17-using-langchain4j-to-analyze-pdf-documents.adoc

```diff
@@ -0,0 +1,197 @@
+public class GoogleGeminiConfig {
+
+    @Produces
+    @ApplicationScoped
+    ChatLanguageModel model() {
+        return GoogleAiGeminiChatModel.builder()
+                .apiKey(System.getenv("GOOGLE_AI_GEMINI_API_KEY"))
+                .modelName("gemini-2.0-flash")
+                .build();
+    }
+}
+----
+
+[NOTE]
+====
+Quarkus LangChain4j will provide autoconfiguration for Gemini in a future release. Currently, manual configuration is required since the Gemini integration is still evolving, with upstream LangChain4j offering three different modules for Google's AI APIs.
+====
+
+=== Define your data structure
+
+Now we need to model the data structure for the lease information that we want the LLM to extract from any lease document.  You can customize these fields based on the information you need from the PDF documents but in our use case below we are extracting the following information:
+
+[source,java]
+----
+public record LeaseReport(
+    LocalDate agreementDate,
+    LocalDate termStartDate,
+    LocalDate termEndDate,
+    LocalDate developmentTermEndDate,
+    String landlordName,
+    String tenantName,
+    String taxParcelId,
+    BigDecimal acres,
+    Boolean exclusiveRights) {
+}
+----
+
+=== Create the REST endpoint
+
+Lastly, we need to create a `LeaseAnalyzerResource` class that will use the LLM to extract the lease information from the PDF document.
+
+[source,java]
+----
+@Inject
+ChatLanguageModel model;
+
+@PUT
+@Consumes(MediaType.MULTIPART_FORM_DATA)
+@Produces(MediaType.TEXT_PLAIN)
+public String upload(@RestForm("file") FileUpload fileUploadRequest) {
+    final String fileName = fileUploadRequest.fileName();
+    log.infof("Uploading file: %s", fileName);
+
+    try {
+        // Convert input stream to byte array for processing
+        byte[] fileBytes = Files.readAllBytes(fileUploadRequest.filePath());
+
+        // Encode PDF content to base64 for transmission
+        String documentEncoded = Base64.getEncoder().encodeToString(fileBytes);
+
+        // Create user message with PDF content for analysis
+        UserMessage userMessage = UserMessage.from(
+                TextContent.from("Analyze the given document"),
+                PdfFileContent.from(documentEncoded, "application/pdf"));
+
+        // Build chat request with JSON response format
+        ChatRequest chatRequest = ChatRequest.builder()
+                .messages(userMessage)
+                .parameters(ChatRequestParameters.builder()
+                        .responseFormat(responseFormatFrom(LeaseReport.class))
+                        .build())
+                .build();
+
+        log.info("Google Gemini analyzing....");
+        long startTime = System.nanoTime();
+        ChatResponse chatResponse = model.chat(chatRequest);
+        long endTime = System.nanoTime();
+        String response = chatResponse.aiMessage().text();
+        log.infof("Google Gemini analyzed in %d ns: %s", (endTime - startTime), response);
```

Comment:
Sorry to nitpick, but can we convert the result to seconds as I think it's a lot for more useful for whoever executes the code.

Reply:
Yep OK seconds makes more sense.

Reply:
Or use thw `java.time` classes for beautification purposes :)